rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,113 @@
+---
+feature: nix-language-version
+start-date: 2022-12-12
+author: @fricklerhandwerk
+co-authors: @thufschmitt @Ericson2314
+shepherd-team: 
+shepherd-leader: 
+related-issues: https://github.com/NixOS/nix/issues/7255
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.
+
+# Motivation
+[motivation]: #motivation
+
+Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.
+This proposal is first step towards overcoming that limitation.
+
+# Detailed design
+[design]: #detailed-design
+
+Introduce a magic comment in the first line of a Nix file:
+
+    #? Nix <version>
+
+where `<version>` is a released version of Nix the given file is intended to work with.
```

Comment:
```suggestion
where `<version>` is a released version of Nix the given file is intended to work with.

Add an appropriate parameter to commands that allow encoding the same information where where files are not involved.
```

Reply:
Maybe one could say that some evaluation contexts (such as a flake with specified language version) could also be used to set the grammar version.

Reply:
See prior art: it doesn't make much sense to do it in a flake, because flakes are written in the same language we're trying to version.

Reply:
The problem with command-line flags is that they affect everything in the evaluation. They affect all imported modules, local or remote. This means that you need a "flag day" of changing the world which is a bad idea. By having a scoped opt-in libraries, or even individual files in a library can opt-in and convert themselves one-by-one which allows gradual ecosystem migration.

Reply:
I don't think you can really avoid semantics. If you are sticking a comment at the top of a file it is just an ordinary comment unless you attach semantics. If you want to achieve something with that comment you need to specify what it means.